File: Target.cpp
#include "Target.h"

#ifdef __APPLE__
#include <GLUT/glut.h>
#else
#include <GL/glut.h>
#endif

Target::Target() {
  center.SetAll(0,10,-80);
  color.SetAll(0.8,0.7,0.8);
  innerRadius = 0.5;
  outerRadius = 4;
  isMoving = false;
  deltaX = 0.02;
}

void Target::SetBBX(double* _bbx) {
  bbx = _bbx;
}

void Target::Update(Node* ballList) {
  if( isMoving ) {
    center.SetX( center.GetX() + deltaX );
    if( center.GetX() < bbx[0] ) {
      center.SetX( bbx[0] );
      deltaX *= -1;
    }
    if( center.GetX() > bbx[1] ) {
      center.SetX( bbx[1] );
      deltaX *= -1;
    }
  }

  bool hit = false;
  bool isLast = false;
  Node* curBall = ballList;
  while(curBall != NULL) {
    Ball& b = curBall->ball;
    Vector3d ballCenter = b.GetCenter();
    if( (ballCenter-center).norm() < outerRadius) {
      if( fabs(ballCenter.GetZ()-center.GetZ()) < 1.0 ) {
	hit = true;
	break;
      }
    }

    curBall = curBall->next;
  }

  if( hit ) {
    //start isMoving
    isMoving = true;
    //change color
    color.SetX( drand48() );
    color.SetY( drand48() );
    color.SetZ( drand48() );
    deltaX *= 1.05;
  }
  
}

void Target::Draw() {

  //void glutSolidTorus(GLdouble innerRadius,
  //                  GLdouble outerRadius,
  //                  GLint nsides, GLint rings);

  glColor3f(color.GetX(),color.GetY(),color.GetZ());
  glPushMatrix();
  glTranslated(center[0],center[1],center[2]);
  glutSolidTorus(innerRadius,outerRadius,20,20);
  glPopMatrix();
}

File: py/Target.py
import Ball
import Vector
import math
import random
from OpenGL.GLUT import *
from OpenGL.GLU import *
from OpenGL.GL import *

class Target:
    def __init__(self):
        # self.center.Set_All(0, 10, -80)
        self.center = Vector.Vector3D(0, 10, -80)
        self.color = Vector.Vector3D(0.8, 0.7, 0.8)
        # self.color.Set_All(0.8, 0.7, 0.8)
        self.innerRadius = 0.5
        self.outerRadius = 4
        self.isMoveing = False
        self.deltaX = 0.02
        self.bbx = None

    def SetBBX(self, _bbx):
        self.bbx = _bbx

    def Update(self, a):
        if self.isMoveing is True:
            self.center.SetX(self.center.GetX() + self.deltaX)
            if self.center.GetX() < self.bbx[0]:
                self.center.SetX(self.bbx[0])
                self.deltaX *= -1
            if self.center.GetX() > self.bbx[1]:
                self.center.SetX(self.bbx[1])
                self.deltaX *= -1
        
        hit = False
        isLast = True
        for i in a:
            ballCenter = i.center
            if (ballCenter - self.center).norm() < self.outerRadius:
                if math.fabs(ballCenter.GetZ() - self.center.GetZ()) < 1.0:
                    hit = True
                    break
            
        if hit is True:
            self.isMoveing = True
            self.color.SetX(random.random())
            self.color.SetY(random.random())
            self.color.SetZ(random.random())
            self.deltaX *= 1.05
        # print("hit is ", hit)
        
    def Draw(self):
        glColor3f(self.color.GetX(),self.color.GetY(),self.color.GetZ())
        glPushMatrix()
        glTranslated(self.center.GetX(), self.center.GetY(), self.center.GetZ())
        glutSolidTorus(self.innerRadius, self.outerRadius, 20, 20)
        glPopMatrix()